File: lib/domain/timeline/timeline_attribute.dart
import 'dart:ui';

import 'package:cobble/domain/timeline/timeline_icon.dart';
import 'package:collection/collection.dart';
import 'package:json_annotation/json_annotation.dart';

import 'game_state.dart';

part 'timeline_attribute.g.dart';

/// Single timeline pin attribute
/// Attributes consist of id, one value (of any of the supported types) and
/// optionally maxLength on array types
///
/// Use factory methods to create attributes with valid values.
@JsonSerializable(includeIfNull: false, createFactory: false)
class TimelineAttribute {
  final int id;
  final String string;
  final List<String> listOfString;
  final int uint8;
  final int uint32;
  final int maxLength;

  TimelineAttribute({
    this.id,
    this.string,
    this.listOfString,
    this.uint8,
    this.uint32,
    this.maxLength,
  });

  Map<String, dynamic> toMap() {
    return _$TimelineAttributeToJson(this);
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is TimelineAttribute &&
          runtimeType == other.runtimeType &&
          id == other.id &&
          string == other.string &&
          ListEquality().equals(listOfString, other.listOfString) &&
          uint8 == other.uint8 &&
          uint32 == other.uint32 &&
          maxLength == other.maxLength;

  @override
  @JsonKey(ignore: true)
  int get hashCode =>
      id.hashCode ^
      string.hashCode ^
      listOfString.hashCode ^
      uint8.hashCode ^
      uint32.hashCode ^
      maxLength.hashCode;

  @override
  String toString() {
    return 'TimelineAttribute{id: $id, string: $string, listOfString: $listOfString, uint8: $uint8, uint32: $uint32, maxLength: $maxLength}';
  }

  static TimelineAttribute title(String title) {
    return TimelineAttribute(
      id: 1,
      string: title,
      maxLength: 64,
    );
  }

  static TimelineAttribute subtitle(String subtitle) {
    return TimelineAttribute(
      id: 2,
      string: subtitle,
      maxLength: 64,
    );
  }

  static TimelineAttribute body(String body) {
    return TimelineAttribute(
      id: 3,
      string: body,
      maxLength: 512,
    );
  }

  static TimelineAttribute tinyIcon(TimelineIcon icon) {
    // Pebble requires most significant bit on all icon numbers to be high.
    // Not sure why it was done that way

    return TimelineAttribute(
      id: 4,
      uint32: icon.toProtocolNumber() | 0x80000000,
    );
  }

  static TimelineAttribute smallIcon(TimelineIcon icon) {
    return TimelineAttribute(
      id: 5,
      uint32: icon.toProtocolNumber() | 0x80000000,
    );
  }

  static TimelineAttribute largeIcon(TimelineIcon icon) {
    return TimelineAttribute(
      id: 6,
      uint32: icon.toProtocolNumber() | 0x80000000,
    );
  }

  static TimelineAttribute ancsAction(int action) {
    return TimelineAttribute(
      id: 7,
      uint32: action,
    );
  }

  static TimelineAttribute cannedResponse(List<String> responses) {
    return TimelineAttribute(
      id: 8,
      listOfString: responses,
      maxLength: 512,
    );
  }

  static TimelineAttribute shortTitle(String shortTitle) {
    return TimelineAttribute(
      id: 9,
      string: shortTitle,
      maxLength: 64,
    );
  }

  static TimelineAttribute locationName(String locationName) {
    return TimelineAttribute(
      id: 11,
      string: locationName,
      maxLength: 64,
    );
  }

  static TimelineAttribute sender(String sender) {
    return TimelineAttribute(
      id: 12,
      string: sender,
      maxLength: 64,
    );
  }

  static TimelineAttribute launchCode(int launchCode) {
    return TimelineAttribute(
      id: 13,
      uint32: launchCode,
    );
  }

  static TimelineAttribute lastUpdated(DateTime time) {
    return TimelineAttribute(
      id: 14,
      uint32: (time.millisecondsSinceEpoch / 1000).round(),
    );
  }

  static TimelineAttribute rankAway(String rank) {
    return TimelineAttribute(
      id: 15,
      string: rank,
    );
  }

  static TimelineAttribute rankHome(String rank) {
    return TimelineAttribute(
      id: 16,
      string: rank,
    );
  }

  static TimelineAttribute nameAway(String name) {
    return TimelineAttribute(
      id: 17,
      string: name,
    );
  }

  static TimelineAttribute nameHome(String name) {
    return TimelineAttribute(
      id: 18,
      string: name,
    );
  }

  static TimelineAttribute recordAway(String record) {
    return TimelineAttribute(
      id: 19,
      string: record,
    );
  }

  static TimelineAttribute recordHome(String record) {
    return TimelineAttribute(
      id: 20,
      string: record,
    );
  }

  static TimelineAttribute scoreAway(String score) {
    return TimelineAttribute(
      id: 21,
      string: score,
    );
  }

  static TimelineAttribute scoreHome(String score) {
    return TimelineAttribute(
      id: 22,
      string: score,
    );
  }

  static TimelineAttribute sportsGameState(TimelineGameState state) {
    return TimelineAttribute(
      id: 23,
      uint8: state.toProtocolNumber(),
    );
  }

  static TimelineAttribute broadcaster(String name) {
    return TimelineAttribute(
      id: 24,
      string: name,
    );
  }

  static TimelineAttribute headings(List<String> headings) {
    return TimelineAttribute(id: 25, listOfString: headings, maxLength: 128);
  }

  static TimelineAttribute paragraphs(List<String> paragraphs) {
    return TimelineAttribute(id: 26, listOfString: paragraphs, maxLength: 1024);
  }

  static TimelineAttribute foregroundColor(Color color) {
    return TimelineAttribute(
      id: 27,
      uint8: color.toProtocolNumber(),
    );
  }

  static TimelineAttribute primaryColor(Color color) {
    return TimelineAttribute(
      id: 28,
      uint8: color.toProtocolNumber(),
    );
  }

  static TimelineAttribute secondaryColor(Color color) {
    return TimelineAttribute(
      id: 29,
      uint8: color.toProtocolNumber(),
    );
  }

  static TimelineAttribute displayRecurring(bool recurring) {
    return TimelineAttribute(
      id: 31,
      uint8: recurring ? 1 : 0,
    );
  }

  static TimelineAttribute shortSubtitle(String shortSubtitle) {
    return TimelineAttribute(
      id: 36,
      string: shortSubtitle,
      maxLength: 64,
    );
  }

  static TimelineAttribute timestamp(DateTime timestamp) {
    return TimelineAttribute(
      id: 37,
      uint32: timestamp.millisecondsSinceEpoch ~/ 1000,
    );
  }

  static TimelineAttribute displayTime(bool displayTime) {
    return TimelineAttribute(
      id: 38,
      uint8: displayTime ? 1 : 0,
    );
  }

  static subtitleTemplateString(String templateString) {
    return TimelineAttribute(id: 47, string: templateString, maxLength: 150);
  }

  static icon(TimelineIcon icon) {
    return TimelineAttribute(
      id: 48,
      uint8: icon.toProtocolNumber(),
    );
  }
}

extension ColorAttribute on Color {
  int toProtocolNumber() {
    return ((alpha ~/ 85) << 6) |
        ((red ~/ 85) << 4) |
        ((green ~/ 85) << 2) |
        (blue ~/ 85);
  }
}

File: android/shared/src/commonMain/kotlin/io/rebble/cobble/shared/data/TimelineAttribute.kt
package io.rebble.cobble.shared.data

import io.rebble.cobble.shared.domain.common.Color
import io.rebble.cobble.shared.domain.common.toProtocolNumber
import io.rebble.cobble.shared.domain.timeline.TimelineGameState
import io.rebble.cobble.shared.domain.timeline.TimelineIcon
import io.rebble.libpebblecommon.packets.blobdb.TimelineItem
import io.rebble.libpebblecommon.structmapper.SUInt
import io.rebble.libpebblecommon.structmapper.StructMapper
import io.rebble.libpebblecommon.util.DataBuffer
import io.rebble.libpebblecommon.util.encodeToByteArrayTrimmed
import kotlinx.serialization.Serializable

@Serializable
@OptIn(ExperimentalUnsignedTypes::class)
data class TimelineAttribute(
        val id: Int,
        val string: String? = null,
        val listOfString: List<String>? = null,
        val uint8: Int? = null,
        val uint32: Long? = null,
        val maxLength: Int = 64
) {
    fun toProtocolAttribute(): TimelineItem.Attribute {
        val content: UByteArray = when {
            string != null -> {
                string.encodeToByteArrayTrimmed(maxLength)
                        .toUByteArray()
            }

            listOfString != null -> {
                val combinedStrings = listOfString.joinToString(separator = "\u0000")
                combinedStrings.encodeToByteArrayTrimmed(maxLength).toUByteArray()
            }

            uint8 != null -> {
                ubyteArrayOf(uint8.toUByte())
            }

            uint32 != null -> {
                SUInt(StructMapper(), uint32.toUInt(), '<').toBytes()
            }

            else -> throw IllegalArgumentException("Received empty timeline attribute: $this")
        }

        return TimelineItem.Attribute(
                id.toUByte(),
                content
        )
    }

    companion object {
        fun fromProtocolAttribute(attr: TimelineItem.Attribute): TimelineAttribute {
            return when (val id = attr.attributeId.get().toInt()) {
                in 1..3, 9, 11, 12, in 15..22, 24 -> {
                    TimelineAttribute(id, attr.content.get().toByteArray().joinToString())
                }

                in 4..7, 13, 14 -> {
                    TimelineAttribute(id, uint32 = DataBuffer(attr.content.get()).getULong().toLong())
                }

                else -> {
                    throw NotImplementedError("Attr ID $id does not have an implemented type.")
                }
            }
        }

        fun title(title: String?) = TimelineAttribute(1, string = title)
        fun subtitle(subtitle: String?) = TimelineAttribute(2, string = subtitle)
        fun body(body: String?) = TimelineAttribute(3, string = body)
        fun tinyIcon(icon: TimelineIcon) = TimelineAttribute(4, uint32 = icon.protocolValue.toLong() or 0x80000000)
        fun smallIcon(icon: TimelineIcon) = TimelineAttribute(5, uint32 = icon.protocolValue.toLong() or 0x80000000)
        fun largeIcon(icon: TimelineIcon) = TimelineAttribute(6, uint32 = icon.protocolValue.toLong() or 0x80000000)
        fun ancsAction(action: Int) = TimelineAttribute(7, uint32 = action.toLong())
        fun cannedResponse(responses: List<String>) = TimelineAttribute(8, listOfString = responses, maxLength = 512)
        fun shortTitle(shortTitle: String) = TimelineAttribute(9, string = shortTitle, maxLength = 64)
        fun locationName(locationName: String) = TimelineAttribute(11, string = locationName, maxLength = 64)
        fun sender(sender: String) = TimelineAttribute(12, string = sender, maxLength = 64)
        fun launchCode(launchCode: Int) = TimelineAttribute(13, uint32 = launchCode.toLong())
        fun lastUpdated(time: Long) = TimelineAttribute(14, uint32 = time)
        fun rankAway(rank: String) = TimelineAttribute(15, string = rank)
        fun rankHome(rank: String) = TimelineAttribute(16, string = rank)
        fun nameAway(name: String) = TimelineAttribute(17, string = name)
        fun nameHome(name: String) = TimelineAttribute(18, string = name)
        fun recordAway(record: String) = TimelineAttribute(19, string = record)
        fun recordHome(record: String) = TimelineAttribute(20, string = record)
        fun scoreAway(score: String) = TimelineAttribute(21, string = score)
        fun scoreHome(score: String) = TimelineAttribute(22, string = score)
        fun sportsGameState(state: TimelineGameState) = TimelineAttribute(23, uint8 = state.protocolValue)
        fun broadcaster(name: String) = TimelineAttribute(24, string = name)
        fun headings(headings: List<String>) = TimelineAttribute(25, listOfString = headings, maxLength = 128)
        fun paragraphs(paragraphs: List<String>) = TimelineAttribute(26, listOfString = paragraphs, maxLength = 1024)
        fun foregroundColor(color: Color) = TimelineAttribute(27, uint8 = color.toProtocolNumber())
        fun primaryColor(color: Color) = TimelineAttribute(28, uint8 = color.toProtocolNumber())
        fun secondaryColor(color: Color) = TimelineAttribute(29, uint8 = color.toProtocolNumber())
        fun displayRecurring(recurring: Boolean) = TimelineAttribute(31, uint8 = if (recurring) 1 else 0)
        fun shortSubtitle(shortSubtitle: String) = TimelineAttribute(36, string = shortSubtitle, maxLength = 64)
        fun timestamp(timestamp: Long) = TimelineAttribute(37, uint32 = timestamp)
        fun displayTime(displayTime: Boolean) = TimelineAttribute(38, uint8 = if (displayTime) 1 else 0)
        fun subtitleTemplateString(templateString: String) = TimelineAttribute(47, string = templateString, maxLength = 150)
        fun icon(icon: TimelineIcon) = TimelineAttribute(48, uint8 = icon.protocolValue)
    }
}